drivers/serial/uart_nrfx_uarte.c

```diff
@@ -710,13 +743,28 @@ static void endrx_isr(struct device *dev)
 
 	data->async->is_in_irq = true;
 
-	if (data->async->rx_next_buf) {
-		nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTRX);
-	}
+	/* ensure rx timer is stopped - it will be restarted in RXSTARTED
+	 * handler if needed
+	 */
 	k_timer_stop(&data->async->rx_timeout_timer);
 
-	size_t rx_len = nrf_uarte_rx_amount_get(uarte)
-			- data->async->rx_offset;
+	/* this is the amount that the EasyDMA controller has copied into the
+	 * buffer
+	 */
+	const int this_amount = nrf_uarte_rx_amount_get(uarte);
+
+	/* it is possible for rx_offset to be more than the length of the
+	 * EasyDMA transfer, as the rx_offset comes from counts of the RXRDY
+	 * event, which occurs before data is transferred via DMA
+	 * So in these cases we clip rx_len
```

Comment:
This comment may be misleading (at least I was a bit confused by it). `rx_offset` can be bigger than `this_amount` because it can also include some data from the next buffer that is already being filled up by the transfer started via the ENDRX_STARTRX shortcut.

Reply:
updated the comment.